Fix decomp for linspace

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147997

In python decompositions, we shouldn't do any non-functional operations for functional operators. This should go away once we start decomposing before functionalization. 

Differential Revision: [D70265200](https://our.internmc.facebook.com/intern/diff/D70265200)

tugsbayasgalan committed:
Fix decomp for linspace

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70265200).

tugsbayasgalan committed:
Update on "Fix decomp for linspace"


In python decompositions, we shouldn't do any non-functional operations for functional operators. This should go away once we start decomposing before functionalization. 

Differential Revision: [D70265200](https://our.internmc.facebook.com/intern/diff/D70265200)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70265200).

bdhirsh commented on torch/_refs/__init__.py:
```diff
@@ -5216,7 +5216,10 @@ def linspace(
+            # we need to convert tensor(0,) to torch.Size([1])
+            if len(start.shape) == 0:
+                return torch.full((steps,), start.item(), dtype=dtype, **factory_kwargs)  # type: ignore[arg-type]
+            return start.clone()
```
the original decomp promised to keep the same metadata as `dtype` and `factory_kwargs`. is that preserved here?

bdhirsh replied:
you probably just want to use `aten.copy(torch.empty(....))` directly to ensure you get the same semantics

(although we should just fix the functionalization-after-decomp story :) )

tugsbayasgalan committed:
Update on "Fix decomp for linspace"


In python decompositions, we shouldn't do any non-functional operations for functional operators. This should go away once we start decomposing before functionalization. 

Differential Revision: [D70265200](https://our.internmc.facebook.com/intern/diff/D70265200)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70265200).

zou3519 reviewed (commented):
seems reasonable but your tests are failing

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70265200).

tugsbayasgalan committed:
Update on "Fix decomp for linspace"


In python decompositions, we shouldn't do any non-functional operations for functional operators. This should go away once we start decomposing before functionalization. 

Differential Revision: [D70265200](https://our.internmc.facebook.com/intern/diff/D70265200)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70265200).

tugsbayasgalan committed:
Update on "Fix decomp for linspace"


In python decompositions, we shouldn't do any non-functional operations for functional operators. This should go away once we start decomposing before functionalization. 

Differential Revision: [D70265200](https://our.internmc.facebook.com/intern/diff/D70265200)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70265200).

tugsbayasgalan committed:
Update on "Fix decomp for linspace"


In python decompositions, we shouldn't do any non-functional operations for functional operators. This should go away once we start decomposing before functionalization. 

Differential Revision: [D70265200](https://our.internmc.facebook.com/intern/diff/D70265200)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70265200).
